### PR TITLE
feat: add imagenet pytorch example

### PR DESCRIPTION
**Pull request by katport**
## Checklist

- [ ] User-facing API changes need the "User-facing API Change" label.
- [ ] Release notes should be added as a separate file under `docs/release-notes/`.
See [Release Note](../docs/release-notes/README.md) for details.
- [ ] Licenses should be included for new code which was copied and/or modified from any external code.

[DET-1234]: https://determinedai.atlassian.net/browse/DET-1234

### CODE REVIEW

File: examples/tutorials/imagenet_pytorch/model_def.py [hunk 1]
```diff
@@ -0,0 +1,159 @@
...
+        self.download_directory = f"/tmp/data-rank{self.context.distributed.get_rank()}"
+        self.data_directory = self.context.get_hparam("data_location")
+
+        optimizer = torch.optim.SGD(self.model.parameters(), self.context.get_hparam("lr"), momentum=self.context.get_hparam("momentum"), weight_decay=self.context.get_hparam("weight_decay"))
+        self.optimizer = self.context.wrap_optimizer(optimizer)
+
+        self.criterion = nn.CrossEntropyLoss()
+
+        lr_sch = torch.optim.lr_scheduler.StepLR(self.optimizer, gamma=.1, step_size=2)
+        self.lr_sch = self.context.wrap_lr_scheduler(lr_sch, step_mode=LRScheduler.StepMode.STEP_EVERY_EPOCH)
```
Comment (naren-determined) [inline examples/tutorials/imagenet_pytorch/model_def.py:56]: Readability-wise, I think it would be easier to not define `lr_sch` and just put this directory into the `self.context.wrap_lr_scheduler`

File: examples/tutorials/imagenet_pytorch/model_def.py [hunk 2]
```diff
@@ -0,0 +1,159 @@
...
+                transforms.Compose([
+                    transforms.RandomResizedCrop(224),
+                    transforms.RandomHorizontalFlip(),
+                    transforms.ToTensor(),
+                    self.normalize,
+                ]))
+
+            return DataLoader(
+                train_dataset, batch_size=self.context.get_per_slot_batch_size(), shuffle=True,
+                num_workers=self.context.get_hparam("workers", pin_memory=True))
```
Comment (naren-determined) [inline examples/tutorials/imagenet_pytorch/model_def.py:76]: Nit: Can you separate each expression/comma out on a different line? Ditto validation.

File: examples/tutorials/imagenet_pytorch/model_def.py [hunk 3]
```diff
@@ -0,0 +1,159 @@
...
+        
+        elif self.context.get_hparam('dataset') == 'cifar':
+
+            transform = transforms.Compose(
+                [transforms.ToTensor(), transforms.Normalize((0.5, 0.5, 0.5), (0.5, 0.5, 0.5))]
+            )
+            trainset = datasets.CIFAR10(
+                root=self.download_directory, train=True, download=True, transform=transform
```
Comment (naren-determined) [inline examples/tutorials/imagenet_pytorch/model_def.py:84]: I'm not a big fan of hardcoding download=True here. I think it would be better to have that as a parameter in the config file that defaults to True just in case users want to change that and pass their own data directory. Ditto validation.

File: examples/tutorials/imagenet_pytorch/model_def.py [hunk 4]
```diff
@@ -0,0 +1,159 @@
...
+        # measure accuracy and record loss
+        acc1, acc5 = self.accuracy(output, target, topk=(1, 5))
+
+        return {"val_loss": loss.item(), "top1": acc1[0], "top5":acc5[0]}
+
+    def adjust_learning_rate(self, optimizer, epoch, args):
+        """Sets the learning rate to the initial LR decayed by 10 every 30 epochs"""
+        lr = args.lr * (0.1 ** (epoch // 30))
+        for param_group in self.optimizer.param_groups:
+            param_group['lr'] = lr
```
Comment (naren-determined) [inline examples/tutorials/imagenet_pytorch/model_def.py:142]: I don't see this being used anywhere. Would it be okay to remove it or perhaps add a comment if it was in the original code?

File: examples/tutorials/imagenet_pytorch/const.yaml [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+  dataset: imagenet
+records_per_epoch: 50000
+searcher:
+  name: single
+  metric: val_loss
+  smaller_is_better: true
+  max_length:
+    epochs: 60
+entrypoint: model_def:ImagenetTrial
+max_restarts: 0
```
[the remark this answers is not in the document]
Comment (katport) [reply examples/tutorials/imagenet_pytorch/const.yaml:24]: This tutorial is getting used for the porting guide. In the porting guide I reference the max_restarts: 0, which is why I've kept it in the yaml files.

File: examples/tutorials/imagenet_pytorch/model_def.py [hunk 6]
```diff
@@ -0,0 +1,159 @@
...
+import torch.nn as nn
+import torchvision.transforms as transforms
+import torchvision.datasets as datasets
+import torchvision.models as models
+
+
+TorchData = Union[Dict[str, torch.Tensor], Sequence[torch.Tensor], torch.Tensor]
+model_names = sorted(name for name in models.__dict__
+    if name.islower() and not name.startswith("__")
+    and callable(models.__dict__[name]))
```
[the remark this answers is not in the document]
Comment (katport) [reply examples/tutorials/imagenet_pytorch/model_def.py:29]: I changed a few things around. This is now removed.

File: examples/tutorials/imagenet_pytorch/model_def.py (hunk 3 above)
Comment (katport) [reply examples/tutorials/imagenet_pytorch/model_def.py:84]: This actually mimics our cifar example where we also set download == True. I've updated it though to at least be an hp flag

**Review by naren-determined — APPROVED** [review]
Lgtm!
